### PR TITLE
Generic scope handler interfaces

### LINKED ISSUE #484: Update scope type modifier to expand on both ends
Right now you can take lines paragraphs and soon non whitespace sequences that expands from both anchor and active of the selection. Currently containing scopes only use start off the selection.

I want both of these to have the same result. (air and bat are in different functions)
`take funk air past bat`
`take air past bat take funk`

I was thinking that we could implement this first thing in the selection/scope modifier by basically calling the specific implementation twice, once for both ends. To day we pass the entire selection to the specific modifier(line, funk, ...). What if we pass in a single position instead?

```js
const startSelection = getSelectionForScope(selection.start, "function"); // function is of course a variable
if (selection.isEmpty() 
  || getNodeAtLocation(selection.start).id === getNodeAtLocation(selection.end).id) {
    return startSelection;
}
const endSelection = getSelectionForScope(selection.end, "function");
return  getSelectionFromPositions(startSelection.start, endSelection.end);
```

- [ ] Add test for https://github.com/pokey/cursorless-vscode/issues/484#issuecomment-1013856057
- [ ] Add test for "funk wrap that" with `that` mark starting in one statement and ending in another (see https://github.com/cursorless-dev/cursorless/issues/883)
- [ ] Add test for "snip funk after that" with `that` mark starting in one function and ending in another (see https://github.com/cursorless-dev/cursorless/issues/883)
- [ ] https://github.com/cursorless-dev/cursorless/issues/901

## _edit (@pokey)_

Let's implement this one in a new `ContainingScope` stage that all containing scopes share.  The algorithm should be as described in https://github.com/cursorless-dev/cursorless/pull/629#issuecomment-1136090441.  The initial implementation will be folded into https://github.com/cursorless-dev/cursorless/pull/629

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
- Implements generic scope handlers
- Adds `"identifier"` scope type
- Closes https://github.com/cursorless-dev/cursorless/issues/797
- Closes #484 

## Checklist

- [x] Add "identifier" scope type
- [x] Switch "word" and "char" to use this new setup, with "identifier" as their parent scope type
- [x] Add test for `"take next word"` when cursor is in the whitespace between two tokens, eg `"foo | barBaz"`
- [x] Add test for `"one token backward"`
- [x] Link issues that this PR closes and clean up related issues
- [x] Remove old modifier stages
- [x] Figure our grand scopes. Just pass ancestorIndex to `getScopesTouchingPosition`
- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [x] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet_html)

### CODE REVIEW

File: src/core/tokenizer.ts [hunk 1]
```diff
@@ -123,10 +123,20 @@ const tokenMatchersForLanguage: Partial<Record<LanguageId, Matcher>> =
...
+  const matcher =
     tokenMatchersForLanguage[languageId as SupportedLanguageId] ??
-    defaultMatcher
-  );
+    defaultMatcher;
+
+  return {
+    ...matcher,
+    // TODO Why is this necessary???????????
+    identifierMatcher: new RegExp(matcher.identifierMatcher),
```
Comment (AndreasArvidsson) [inline src/core/tokenizer.ts:133]: I can't for the life of me understand why I need to create a new instance of this regex. As far as I can tell the old and new one are identical. Could we be running into a concurrency problem?

File: src/processTargets/getModifierStage.ts [hunk 2]
```diff
@@ -63,8 +39,9 @@ export default (modifier: Modifier): ModifierStage => {
     case "trailing":
       return new TrailingStage(modifier);
     case "containingScope":
```
Comment (AndreasArvidsson) [inline src/processTargets/getModifierStage.ts:41]: Noticed that containing and every are two separate modifiers. The different scope types are not modifiers per se.

File: src/processTargets/modifiers/ContainingScopeStage.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+      startTarget.createContinuousRangeTarget(
+        target.isReversed,
+        endTarget,
+        true,
+        true
+      ),
+    ];
+  }
+
+  private runLegacy(
```
Comment (AndreasArvidsson) [inline src/processTargets/modifiers/ContainingScopeStage.ts:75]: This should of course be removed ones all the scope types are migrated into scope type handlers

File: src/processTargets/modifiers/scopeHandlers/TokenScopeHandler.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      (contentRange) =>
+        new TokenTarget({
+          editor,
+          isReversed,
+          contentRange,
+        })
+    );
+  }
+
+  protected getContainingIndicesForPosition(
```
Comment (AndreasArvidsson) [inline src/processTargets/modifiers/scopeHandlers/TokenScopeHandler.ts:35]: Notice how token overwrites the base class behavior for containing scope for a position(empty range)

File: src/processTargets/modifiers/scopeHandlers/scopeHandler.types.ts [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+      const intersection = r.intersection(iterationScope);
+      return intersection != null && !intersection.isEmpty;
+    });
+  }
+
+  protected abstract getEveryTarget(
+    editor: TextEditor,
+    contentRange: Range,
+    isReversed: boolean,
+    hasExplicitRange: boolean
```
Comment (pokey) [inline src/processTargets/modifiers/scopeHandlers/scopeHandler.types.ts:192]: I'm not sure this is what we want to be passing in here; won't "next" want to fake it to `false`?  I think it's confusing to have something that is an attribute of the target and then to hack it to be false.  Could we maybe just name the variable something else?

I think it would be useful to see "next" implemented, as this impl is still doing the hacky faked explicit range thing via calling "every".  Shouldn't be too hard, no?

Comment (AndreasArvidsson) [reply src/processTargets/modifiers/scopeHandlers/scopeHandler.types.ts:192]: Yeah probably. As I said at the top names could probably be improved. 

I can do that. I'm just a bit has content of having these different implementations: new/legacy without support from the type system.

Comment (AndreasArvidsson) [reply src/processTargets/modifiers/scopeHandlers/scopeHandler.types.ts:192]: Ok now relative and ordinal scope stages use the new scope handler for token

File: src/processTargets/modifiers/scopeHandlers/TokenScopeHandler.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  ): Target[] {
+    const tokenRanges = getTokensInRange(
+      editor,
+      expandToFullLine(editor, contentRange)
+    ).map(({ range }) => range);
+
+    const filteredTokenRanges = hasExplicitRange
+      ? this.filterRangesByIterationScope(contentRange, tokenRanges)
+      : tokenRanges;
+
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply src/processTargets/modifiers/scopeHandlers/TokenScopeHandler.ts:24]: Maybe` but I'm not sure if that is going to be consistent across all different scope types.

[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply src/processTargets/modifiers/scopeHandlers/TokenScopeHandler.ts:24]: We might have multiple iteration scopes. Functions inside/outside classes for example. I'm not sure it will differ but it night since we haven't thought about that implementation yet it's hard to say.

Also this line is in the middle of another function I'm unsure about how much code we save by extracting this.

File: src/processTargets/modifiers/scopeHandlers/TokenScopeHandler.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    return filteredTokenRanges.map(
+      (contentRange) =>
+        new TokenTarget({
+          editor,
+          isReversed,
+          contentRange,
+        })
```
Comment (pokey) [inline src/processTargets/modifiers/scopeHandlers/TokenScopeHandler.ts:31]: I'd be tempted to have this construction be a separate function in the scope handler.  Then we can avoid having to construct targets that will end up getting filtered out when the above filter step is in the base class.  Idk; I don't feel super strongly on this one tho. Might be simpler to just have one function that returns targets rather than one that returns the ranges and another that constructs the targets

One thought though is that returning something other than a target would allow us to add domains to each range, so that the base class can use that for intersection

Comment (AndreasArvidsson) [reply src/processTargets/modifiers/scopeHandlers/TokenScopeHandler.ts:31]: I've had that thought as well. The problem is that the constructor for some of these targets has a lot more information than this. Basically this is the untyped clone parameter issue all over again.

Comment (pokey) [reply src/processTargets/modifiers/scopeHandlers/TokenScopeHandler.ts:31]: Another advantage of separating the construction is that we can remove another arg from this function signature; don't need `isReversed` here

Would be kinda cool if this function just took an editor and a range

Comment (AndreasArvidsson) [reply src/processTargets/modifiers/scopeHandlers/TokenScopeHandler.ts:31]: Agreed. If we can just find a good enough solution for the constructor parameters then I'm totally with you.

File: src/processTargets/modifiers/scopeHandlers/TokenScopeHandler.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (pokey) [reply src/processTargets/modifiers/scopeHandlers/TokenScopeHandler.ts:24]: Also allows us to make `filterRangesByIterationScope` private rather than protected, which is nice.  Just trying to simplify the contract between base class and scope class as much as possible

[the remark this answers is not in the document]
Comment (pokey) [reply src/processTargets/modifiers/scopeHandlers/TokenScopeHandler.ts:24]: I know you view it as an implementation detail, but I think it should be a first-class api.  For example, we could imagine third parties might implement their own scope types via extensions.  I know that sounds far out, but the more we can view separate components of cursorless as independent pieces with well-defined api contracts, the better

[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply src/processTargets/modifiers/scopeHandlers/TokenScopeHandler.ts:24]: I will give it a try

[the remark this answers is not in the document]
Comment (pokey) [reply src/processTargets/modifiers/scopeHandlers/TokenScopeHandler.ts:24]: Looking at the code more, I'm actually tempted to move the filter stage into the modifiers.  It's only "every" and ordinal that need to do this; containing and relative don't filter

If you're really concerned about the duplication, we could revert to having ordinal call every stage; I actually don't think that's too crazy

[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply src/processTargets/modifiers/scopeHandlers/TokenScopeHandler.ts]: Updated now. Definitely needs to have a look at the typing system, but something general is in place now at least.

**Comment by pokey** [comment]
Ok super helpful to see the code written out; great call to just bang it out.

Here's what I'm thinking.  Scope type exposes two functions:

- A function that takes a position and returns containing iteration range, just as a range
- A function that takes an iteration range and returns a list of scopes

I think it would look as follows

```ts
interface ScopeDefinition {
   getIterationRange(editor: TextEditor, position: Position): Range;
   getScopesInIterationRange(editor: TextEditor, range: Range): Scope[];
}

interface Scope {
   domain: Range;
   getTarget(isBackward: boolean): Target;
   getPriority?(): number;
}
```

Then the base class / intermediary has an interface as follows:

```ts
/** For use as return type of {@link ScopeHandler.getTargetsForRange} */
interface IterationScope {
   targets: Target[]
   containingStartIndex: number;
   containingEndIndex: number;
}

class ScopeHandler {
   getTargetsForRange(editor: TextEditor, range: Range, isBackward: boolean): IterationScope {
      const startIterationRange = self.scopeDefinition.getIterationRange(editor, range.start)
      const iterationRange = contains(startIterationRange, range.end) ? startIterationRange : startIterationRange.union(self.scopeDefinition.getIterationRange(editor, range.end)

      const scopes = self.scopeDefinition.getScopesInRange(editor, iterationRange)

      // Then compute containing scope by intersection with domain for non-empty range
      // For empty range, pick weakly containing domain, breaking ties using `scope.getPriority()` if it exists, else prefer right
      // Then call `getTarget()` on the ones we want to return for `targets`
   }
}
```

Then

- containing scope modifier stage is implemented by grabbing the `iterationScope.containingIndices` and constructing range target
- `every` scope modifier stage is implemented by grabbing the `iterationScope.targets` and just doing an intersection internally if `hasExplicitRange`
- `ordinal` scope modifier stage is implemented by calling the `every` stage, like we do today.  I think it is actually good that there is a hard dependency here, because we want `every` and `ordinal` to behave identically
- `relative` scope modifier stage is implemented by looking at containing scope and list of targets and doing its magic from there


Note that the only place that we ever refer to `hasExplicitRange` is in the `every` scope modifier stage

wdyt?

**Comment by pokey** [comment]
One other question is how we handle "every round". I guess maybe we just have it return an empty range for getIterationRange? I'm not sure we want to do any expansion

Then for getIterationScope, if any pairs start or end inside the iteration range, it returns all top-level pairs that it it intersects with. Top-level here means not contained by any other pair intersecting with iteration range.

If no pairs start or end inside the iteration range, we return the smallest pair containing the iteration range

So I think this works within the api I defined above; just wanted to write it out to see

**Comment by pokey** [comment]
For functions, we expand to class in getIterationRange

In getScopesInRange, I think we just check that the range is just one class, and return its functions if so? Otherwise error? Idk

File: src/core/tokenizer.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply src/core/tokenizer.ts:133]: @pokey Any ideas?

Comment (pokey) [reply src/core/tokenizer.ts:133]: What happens if you don't do that?

Comment (AndreasArvidsson) [reply src/core/tokenizer.ts:133]: It will fail to match in some of the test cases and we get no identifiers back causing the entire test to fail.

**Comment by pokey** [comment]
Ok what should the following do?

```
aaa bbb|. ccc
```

Cursor is the `|`; saying `"two tokens"`?  Should it refer to `bbb.` or `. ccc`?  I think probably the consistent thing is `bbb.`, because it's two tokens starting from `"token"`.  One could also argue it should always just look to the right, though that breaks down when the cursor is in the middle of a token, which should obviously be that token and the following.

Unfortunately, I don't think it's as simple as `containing` and then adding as many scopes as necessary.  For example:

```
{
    aaa {bbb} ccc {ddd} eee
    {fff}
}
```

In this case, if the user says `"three curlies line air"`,  I think it should be `{bbb}` past `{fff}`, but if we treat `"<number> <scope>s"` as first expanding to containing, it will expand to the big curly containing everything and then have nothing to add

So my thinking is as follows:

If the input target is empty, then index 0 should refer to containing scope.  If the input target is not empty, then index 0 should refer to all intersecting scopes.  Wdyt?

**Comment by pokey** [comment]
@AndreasArvidsson fyi https://github.com/cursorless-dev/cursorless/pull/1031/commits/969bc720e98b89a7526656622b40375285b9c5f7.  I accidentally imported the legacy v2 `ScopeType` in a couple places when I did auto-import.  Shall we make it a policy not to export any legacy types without a `vN` suffix?

**Review by pokey — COMMENTED** [review]
Ok I just need to add some jsdocs / comments to `RelativeScopeStage` but otherwise I think this is good to go.  Feel free to have a peek before meet-up if you get a minute

File: src/test/suite/fixtures/recorded/relativeScopes/clearTwoTokenBackward2.yml [hunk 8]
```diff
@@ -18,9 +18,5 @@ initialState:
     - anchor: {line: 0, character: 8}
       active: {line: 0, character: 8}
   marks: {}
-finalState:
-  documentContents: "  ccc"
-  selections:
-    - anchor: {line: 0, character: 0}
-      active: {line: 0, character: 0}
 fullTargets: [{type: primitive, mark: {type: cursor}, modifiers: [{type: relativeScope, scopeType: {type: token}, offset: 0, length: 2, direction: backward}]}]
+thrownError: {name: NoContainingScopeError}
```
[the remark this answers is not in the document]
Comment (pokey) [reply src/test/suite/fixtures/recorded/relativeScopes/clearTwoTokenBackward2.yml:22]: Update from meet-up: agreed

File: src/processTargets/modifiers/scopeHandlers/LineScopeHandler.ts [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+import type { IterationScope, TargetScope } from "./scope.types";
+import type { ScopeHandler } from "./scopeHandler.types";
+
+export default class LineScopeHandler implements ScopeHandler {
+  public readonly scopeType: ScopeType = { type: "line" };
+  public readonly iterationScopeType: ScopeType = { type: "document" };
+
+  constructor(_scopeType: ScopeType, _languageId: string) {
+    // empty
+  }
```
[the remark this answers is not in the document]
Comment (pokey) [reply src/processTargets/modifiers/scopeHandlers/LineScopeHandler.ts:17]: Decision: Make them optional

**Comment by pokey** [comment]
fwiw this is what it looks like to add a scope type now: https://github.com/cursorless-dev/cursorless/pull/1031/commits/08188162a8e5b40c96dbc3c55ebb49008b606806 cc/ @AndreasArvidsson

**Review by pokey — APPROVED** [review]
Ok I think it's good to go!

**Comment by pokey** [comment]
Filed #1052 to track migrating remaining scope types
